### ISSUE TITLE
:otter: NixOS 23.05 — Feature Freeze & Release Blockers

### ISSUE THREAD

**Issue by RaitoBezarius** [NixOS/nixpkgs#224457]
Hello everyone, we are [Ryan Lahfa](https://github.com/RaitoBezarius) & [Martin Weinelt](https://github.com/hexa), the release managers for NixOS 23.05 :otter: ("_Stoat_").

As we approach the next release of NixOS, it's time to discuss the upcoming feature freeze period. We want to ensure that this release is as stable and reliable as possible, and your contributions are crucial to achieving that goal, here are the two upcoming dates in that context:

- **17th April 2023 (17/04/2023)**: Restrict breaking changes to [Release Critical Packages](https://nixos.github.io/release-wiki/Release-Critical-Packages.html)
- **1st May 2023 (01/05/2023)**: Restrict all breaking changes except for desktop environments.

The complete timeline is available here: 

- https://github.com/NixOS/nixpkgs/issues/223562

The first restriction will be in place in two weeks, therefore, we encourage all teams to check in now and mention possible roadblocks, so we get a good overview of what's ahead.

# Blockers

Whether you were pinged or not, we encourage everyone to create issues for release blockers and add them to the [23.05 Blockers project](https://github.com/orgs/NixOS/projects/14).

# Check-in

Please comment your status quo and possible issues as soon as possible. We'll check teams once they have checked in.

If you think some subsystem, person or team should be added or removed from this list, update `maintainers/team-list.nix` in time before the next feature freeze announcement.

# Desktop environments

- [x] Deepin: @wineee
- [x] Cinnamon: @bobby285271 @mkg20001
- [ ] LXQt: @NixOS/lxqt 
- [ ] Xfce: @romildo @muscaln
- [ ] MATE: @johannesloetzsch @romildo
- [ ] Enlightenment: @NixOS/enlightenment 
- [x] GNOME: @NixOS/gnome 
- [x] Qt / KDE: @NixOS/qt-kde 
- [ ] Lumina: @NixOS/lumina
- [x] Pantheon: @NixOS/pantheon 

# Platforms

- [ ] Darwin: @NixOS/darwin-maintainers 
- [x] Mobile: @NixOS/mobile-nixos 

# Languages ecosystem

- [x] BEAM: @NixOS/beam 
- [ ] C: @matthewbauer @Mic92
- [x] Coq: @CohenCyril @Zimmi48 @siraben @vbgl @Alizter
- [x] Dhall: @Gabriella439 @ehmry
- [x] Go: @c00w @kalbasit @Mic92 @zowoq @qbit
- [x] Haskell: Haskell: @NixOS/haskell 
- [x] Lua: @NixOS/lua
- [ ] Perl: @stigtsp
- [x] PHP: @NixOS/php
- [x] Python: @FRidh @mweinelt @jonringer
- [x] R: @bcdarwin @jbedo
- [ ] Ruby: @marsam
- [x] Rust: @figsoda @LnL7 @Mic92 @tjni @winterqt @zowoq
- [x] Node.js: @lilyinstarlight @marsam @winterqt @yu-re-ka
- [ ] Crystal: @NixOS/crystal-lang 
- [x] OCaml: @NixOS/ocaml 

# Compilers, build systems

- [ ] Bazel: @mboes @marsam @uri-canva @avdv @olebedev @groodt @aherrmann @ylecornec
- [x] LLVM: @dtzWill @Ericson2314 @lovek323 @primeos @alyssais @RaitoBezarius @rrbutani @sternenseemann
- [x] CUDA: @NixOS/cuda-maintainers 
- [x] ROCm: @NixOS/rocm-maintainers 
- [ ] Exotic platforms: @NixOS/exotic-platform-maintainers 
- [ ] Static builds: @NixOS/static 

# Programs

- [ ] Freedesktop: @NixOS/freedesktop 
- [x] Geospatial: @NixOS/geospatial 
- [ ] Kodi: @NixOS/kodi 
- [x] Podman: @NixOS/podman 
- [x] Kubernetes: @NixOS/kubernetes 
- [x] NeoVim: @NixOS/neovim 
- [x] emacs: @NixOS/emacs 

# Modules

- [x] ACME: @aanderse @andrew-d @arianvp @emilazy @flokli @m1cr0man
- [x] systemd: @NixOS/systemd 

# Installer

- [ ] Calamares NixOS extensions: @NixOS/calamares-nixos-extensions 

# Internals

- [x] Docs: @asymmetric @ryantm
- [x] NixOS Modules / internals: @Ericson2314 @infinisil @alyssais @roberth
- [ ] Nix/nix-cli ecosystem: @Profpatsch @edolstra @grahamc @nbp
- [x] NixOS tests: @tfc
- [x] Marketing: @garbas @tomberek

# Finally

No issue is too big or too small, but let's remember that we are all working on the project voluntarily in our free time here, so let's focus on issues that can be realistically addressed in the remaining time before the release.

We thank everyone for their contribution!

**Comment by bobby285271** [comment]
- **Budgie:** Though there is no release freeze ping team for this, I want to mention that [Budgie desktop](//blog.buddiesofbudgie.org/) is now [available](//github.com/NixOS/nixpkgs/pull/216737) thanks to `@FedericoSchonborn`'s work.

- **Cinnamon:** No blockers. We will likely ship 5.6 for 23.05 unfortunately since their release schedule [clashes with ours](//github.com/NixOS/nixpkgs/pull/201328#issuecomment-1327006166), otherwise we are in good state. Note that cjs 5.8 will support mozjs 102 but [cjs 5.6 still uses 78](//github.com/NixOS/nixpkgs/issues/157874).

- **GNOME:** I am trying to list some of the blockers I can think of [here](//github.com/NixOS/nixpkgs/projects/53#column-19392674) to get [GNOME 44 PR](//github.com/NixOS/nixpkgs/pull/218143) merged, though `@jtojnar` should know the situation better. Ideally we can also land 44.1, which is scheduled on [April 22](//wiki.gnome.org/FortyFour).

- **Pantheon:** No blockers. We are now using mutter 43 (in [#218143](//github.com/NixOS/nixpkgs/pull/218143)) and all packages are up to date. I am aware of regressions in multitasking view introduced by [mutter](//github.com/elementary/gala/issues/1519) and [glib](//github.com/elementary/gala/issues/1580) update and will apply fixes once available.

**Comment by DianaOlympos** [comment]
BEAM: we have nothing specific to this release. OTP 26 will probably drop after the cutoff, but that is not a problem. There is probably a critical patch coming through, for all last 3 versions, but it should be through before the freeze. Not security, just a nasty bug fix. Not announced yet, don't ask how I know about it. But we will backport it anyway if need be.

**Comment by RaitoBezarius** [comment]
Missed pings due to GitHub limits:

# Modules

- [ ] ACME: @emilazy @flokli @m1cr0man

# Internals

- [ ] Docs: @asymmetric @ryantm
- [ ] NixOS Modules / internals:  @infinisil @roberth
- [ ] Nix/nix-cli ecosystem: @Profpatsch @edolstra @grahamc @nbp
- [ ] NixOS tests: @tfc
- [ ] Marketing: @garbas @tomberek

**Comment by tfc** [comment]
@RaitoBezarius no blockers on the NixOS tests front.

**Comment by RaitoBezarius** [comment]
**LLVM** : AFAIK, we are still working towards ~four things:

- [ ] Remove unneeded LLVM versions: https://github.com/NixOS/nixpkgs/issues/44372
- [ ] Unify llvmPackages_15 and llvmPackages_git code:
   - https://github.com/NixOS/nixpkgs/pull/222901
   - https://github.com/NixOS/nixpkgs/pull/222902
   - https://github.com/NixOS/nixpkgs/pull/222903
   - https://github.com/NixOS/nixpkgs/pull/222894
- [ ] Introduce llvmPackages_16: https://github.com/NixOS/nixpkgs/pull/223282 -- needed for next version of Chromium M113, planned for 2nd of May, already in beta.
- [ ] Upgrade `latest` to 15: https://github.com/NixOS/nixpkgs/pull/213202

Of which, 1 is really a blocker: `llvmPackages_16` introduction.

**Comment by mweinelt** [comment]
We don't have any blockers in python currently.

**Comment by infinisil** [comment]
No blockers for the NixOS Module System and its internals

**Comment by ulrikstrid** [comment]
Not a blocker for OCaml, but I would be very happy if we could move over to Dune 3 more and remove `useDune2`.

**Comment by Gabriella439** [comment]
No blockers for Dhall

**Comment by RaitoBezarius** [comment]
> Not a blocker for OCaml, but I would be very happy if we could move over to Dune 3 more and remove `useDune2`.

Don't hesitate to open an "epic issue" and attach it to this board: https://github.com/NixOS/nixpkgs/projects/18 to track the progress of this cleanup.

**Comment by rnhmjoj** [comment]
It's not my area, but https://github.com/NixOS/nixpkgs/issues/208412 should be a blocker.

**Comment by etu** [comment]
As for PHP, the only thing we *may* consider a blocker is this (removal of PHP 8.0): https://github.com/NixOS/nixpkgs/issues/224505

However, it's not super critical. Removing PHP 8.0 would be *nice* because the support for it from upstream will give it's last patches at end of November and then after that's it's EOL. So the supported versions will also get that same patches at the end of that month. And by that time, we should be about ready with a new release anyways and when the next set of PHP patches comes out (end of December), NixOS 23.05 is EOL anyways.

**Comment by SomeoneSerge** [comment]
**cuda**: it doesn't seem like we have any blockers. We have been experiencing certain issues with libstdc++ after the gcc11->gcc12 update, and they certainly must be solved before the release, but I think we're almost done with them. I'll edit the task list in this post when we have updates

- [ ] https://github.com/NixOS/nixpkgs/issues/220341

**Comment by imincik** [comment]
**Geospatial**: QGIS needs update from 3.28.3 to 3.30.1 and QGIS LTS needs update from 3.22.16 to 3.28.5 (cc @lsix @sikmir @willcohen)

**Comment by SomeoneSerge** [comment]
Tensorflow, which is a dependency for so many things, is being marked [insecure](https://github.com/NixOS/nixpkgs/pull/224436) and needs to be updated lest hydra stop building it. This is a blocker because users can't afford to just build tensorflow locally.

Unfortunately, tensorflow is too big and complex to be handled by the automatic python updates offered by the Python maintainers (such automatic updates tend to result in a cascade of breakages), and it's not quite in scope of CUDA maintainers either.

In other words, it would seem that we lack a team that could take responsibility of the machine learning and scientific computing infrastructure, and own tensorflow (opencv4, pytorch, jax, ...) in particular

- [ ] https://github.com/NixOS/nixpkgs/issues/224353

Kudos to @mweinelt for bringing this up on matrix

**Comment by sternenseemann** [comment]
**Haskell**: https://github.com/NixOS/cabal2nix/pull/597 (and follow up support for `CABAL_DIR`) should be merged, released and propagate to nixpkgs. Can be backported in a pinch though.

*  `haskell.packages.ghc96.haskell-language-server` should work on release. I believe this is already done on `haskell-updates`.
*  As I understand it, there is [some mess](https://github.com/NixOS/nixpkgs/pull/223827#issuecomment-1493428275) regarding cachix 1.4 at the moment. @roberth and @domenkozar seem to be working on it already. If a solution doesn't present itself, it may be prudent to downgrade to cachix 1.3 to begin with. Since cachix falls under the online service exception, breaking changes can later be backported.
* We should make [interaction with stack's official bindists](https://github.com/NixOS/nixpkgs/issues/223390) work more smoothly. @cdepillabout will look into it.
* Get https://github.com/NixOS/nixpkgs/pull/224542 or an equivalent change merged.

Things I want to advertise for is mainly someone looking into https://github.com/NixOS/nixpkgs/issues/208812, since I myself am unlikely to. Would be a shame if a fix doesn't present itself, as it'd be a regression from 22.11.

**Comment by mweinelt** [comment]
Our Ampere Altra boxes don't boot with the 6.1 LTS kernels. Was seen on the aarch64 community builder, but I expect this to also be an issue for our build farm.

https://github.com/nix-community/aarch64-build-box/pull/152

cc @cole-h @grahamc @Ma27 https://github.com/NixOS/nixpkgs/pull/215313

**Comment by garbas** [comment]
>     * [ ]  Marketing: @garbas @tomberek

Regarding the marketing stuff, where could we help you out?

Here is our current plan:
- A banner on the front page of course.
- We will share the release (few times over the month after the release) with on social media accounts.
- We are resurrecting the newsletter and we'll put the 23.05 release into a spotlight there as well.
- Would love to held a release preparation meeting on live.nixos.org, if there is interest from the NixOS team of course. The idea is to bring some transparency to the process, invite users to ask questions, discuss some PRs, ... anything you want actually, NixOS team is in charge of the agenda.

What we would need is some help with to list what is new in 23.05. @RaitoBezarius and @mweinelt (and the rest of NixOS team) would you have time to jump on call to discuss this and if needed write it down together (maybe you already started writing the highlights)?

What we don't have the capacity to do so (yet) but would love to do if somebody :
- Produce some nice "What's new in NixOS 23.05?" video
- Contact the usual portals to feature us
- There are other ideas (Show how many have upgraded to 23.05, ...) but somebody always needs to do it. If you have some time to spread the NixOS please ping us in NixOS Marketing channel (#marketing:nixos.org) to coordinate and to help you out.

**Comment by m1cr0man** [comment]
No blockers for ACME 🙂

**Comment by samueldr** [comment]
Otter than https://github.com/NixOS/nixpkgs/pull/224421, nothing problematic I can think of for Mobile NixOS.

**Comment by ElvishJerricco** [comment]
I think this would be a good time to un-hide the `boot.initrd.systemd` options from the manual. For one, apparently we haven't been all that good about hiding them so a lot of them are visible anyway. But I also think we've reached the point that it's very nearly feature complete and a lot of people are using it on their systems. It would be good to have it in the documentation for this release so we can potentially start thinking about making it default in the next year or so.

**Comment by Flakebi** [comment]
No blockers for **ROCm**

**Comment by ryantm** [comment]
I'm not aware of blockers for documentation. @pennae has been doing a lot of work on manual rendering and might have something to say about the state of that.

**Comment by bcdarwin** [comment]
No blockers for R that I'm aware of 

@jbedo

**Comment by ulrikstrid** [comment]
https://github.com/NixOS/nixpkgs/issues/224569 should probably be a blocker drive or breaks scanning currently.

**Comment by Et7f3** [comment]
> In other words, it would seem that we lack a team that could take responsibility of the machine learning and scientific computing infrastructure, and own tensorflow (opencv4, pytorch, jax, ...) in particular

PyTorch seems broken on Darwin https://github.com/NixOS/nixpkgs/pull/225441 maybe this failure is only for intel.

**Comment by roberth** [comment]
@RaitoBezarius 
> We should probably have a Nix team in nixpkgs.

What's you reasoning? Maybe open a separate thread if you think we have an organizational problem around Nix in Nixpkgs.

Let's keep this thread on the topic of the NixOS release.

**Comment by teto** [comment]
We are working to get neovim 0.9 in the release https://github.com/NixOS/nixpkgs/pull/225152, should be fine.
I would also like to enable structuredAttrs in the lua ecosystem to ease future backports https://github.com/NixOS/nixpkgs/pull/224553

**Comment by wineee** [comment]
***DDE***

Work on adapting `Deepin Desktop Environment` is still in progress，there are currently 3 necessary PRs remaining to be processed：

- https://github.com/NixOS/nixpkgs/pull/225400 
- https://github.com/NixOS/nixpkgs/pull/225534
- https://github.com/NixOS/nixpkgs/pull/215075 
- [TODO] A nixos modules of dde, [basically from this](https://github.com/linuxdeepin/dde-nixos/blob/main/flake.nix)

The DDE component version is consistent with deepin v20 currently，but there are two exceptions：
1. File Manager:
dde-file-manager  requires really heavy patching ([like this](https://github.com/linuxdeepin/dde-nixos/blob/dev/packages/apps/dde-file-manager/default.nix)),  and the v23 version seems to work fine on v20 base

2.  Window Manager:
deepin-kwin in v20 is just to old, can‘t  compile at all.  But embarrassingly, we don't have a tag that can be used normally. In  5.25.0, upstream removed deepin-wm-dbus，which moved to dde-appearance(One of dde-daemon's successors on v23).

Porting DDE v23 is still need months away，It's not stable even on deepin.

**Comment by flokli** [comment]
https://github.com/NixOS/nixpkgs/pull/226192 should probably get merged, so people can't invoke `hostnamectl`/`timedatectl` manually (regression introduced with https://github.com/NixOS/nixpkgs/pull/216826).

**Comment by ElvishJerricco** [comment]
> I think this would be a good time to un-hide the `boot.initrd.systemd` options from the manual. For one, apparently we haven't been all that good about hiding them so a lot of them are visible anyway. But I also think we've reached the point that it's very nearly feature complete and a lot of people are using it on their systems. It would be good to have it in the documentation for this release so we can potentially start thinking about making it default in the next year or so.

#226237

**Comment by RaitoBezarius** [comment]
https://github.com/NixOS/nixpkgs/issues/227431 is a blocker, I advise to try https://github.com/medhefgo/systemd/commit/3ed1d966f00b002ed822ca9de116252bd91fe6c3 and for people who have too much time in life, to patch OVMF to cripple their EFI read API, reproduce the bug and test the fix properly with a NixOS test.

**Comment by srhb** [comment]
The upgrade of Nix to 2.12 has a breaking (unintentional?) change to the language regarding JSON serialization of floats. I don't know whether or not it's a dealbreaker for NixOS 23.05, but I wanted to highlight it regardless.

https://github.com/NixOS/nix/issues/8259

**Comment by kira-bruneau** [comment]
I have a PR that has a breaking change in an a package, but not a NixOS module: #229321

Should we wait until the next release to get this update in?

**Comment by euank** [comment]
We've had discussion about handling k3s and version support over in #224483.

The outcome of that is, I believe, that we don't want to include k3s versions in a release which will go out of support while that release is active.

I've opened a PR to drop those versions over here: https://github.com/NixOS/nixpkgs/pull/233625

I'm not 100% if this issue is the correct place to highlight it since this issue seems like it could be "blockers for cutting the initial release branch", while this one is closer to a "blocker for cutting the final release". This is our first time for this one, so please let me know if there's a more appropriate place for this.

**Comment by RaitoBezarius** [comment]
NixOS 23.05 has been released: https://discourse.nixos.org/t/nixos-23-05-released/28649 ! Thank you to everyone who answered here and helped us to drive this process. See you at the next release!
